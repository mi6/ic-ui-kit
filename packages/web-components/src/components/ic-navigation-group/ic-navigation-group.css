@import "../../global/normalize.css";

/**
 * @prop --ic-z-index-navigation-item: z-index of navigation group item
 */

:host {
  display: block;
}

:host(.in-side-menu) {
  border-bottom: var(--ic-border-light);
  padding: var(--ic-space-md) 0;
}

:host .navigation-group {
  height: 100%;
  width: fit-content;
  color: var(--ic-theme-text);
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0 var(--ic-space-md);
  transition: var(--ic-easing-transition-slow);
  position: relative;
  background: none;
  border: none;
}

:host(.in-side-menu) .navigation-group {
  height: 2.5rem;
  width: 100%;
  text-align: left;
}

:host(.in-side-menu) .navigation-group .ic-typography-label {
  width: 11.875rem;
}

:host(.in-side-menu) .navigation-group-side-menu {
  color: var(--ic-action-default);
  justify-content: flex-start;
  padding: 0 var(--ic-space-md) 0 var(--ic-space-md);
}

:host(.in-side-menu) .navigation-group-side-menu-collapsed,
:host(.in-side-menu) .navigation-group-side-menu-expanded {
  color: var(--ic-action-default);
  justify-content: flex-start;
  padding-left: var(--ic-space-md);
}

:host(.in-side-menu) .navigation-group-side-menu-expanded {
  margin-bottom: var(--ic-space-xs);
}

:host(.in-side-menu) .navigation-group-side-menu-collapsed::after,
:host(.in-side-menu) .navigation-group-side-menu-expanded::after {
  border-style: solid;
  border-width: 0.125em 0.125em 0 0;
  content: "";
  display: inline-block;
  height: 0.45em;
  left: 0.75em;
  position: relative;
  top: -0.1em;
  transform: rotate(135deg);
  vertical-align: top;
  width: 0.45em;
}

:host(.in-side-menu) .navigation-group-side-menu-expanded::after {
  top: 0.15em;
  transform: rotate(-45deg);
}

:host .navigation-group:hover,
:host .navigation-group:active,
:host .selected {
  background-color: var(--ic-architectural-20);
  color: var(--ic-color-primary-text);
  outline: none;
  cursor: pointer;
}

:host(:not(.in-side-menu)) .navigation-group:focus {
  box-shadow: var(--ic-border-focus);
  border-radius: var(--ic-border-radius);
  outline: var(--ic-hc-focus-outline);
  z-index: var(--ic-z-index-navigation-item);
  transition: box-shadow var(--ic-easing-transition-fast);
}

:host(.in-side-menu) .navigation-group-side-menu:hover,
:host(.in-side-menu) .navigation-group-side-menu:active,
:host(.in-side-menu) .navigation-group-side-menu:focus {
  background-color: var(--ic-architectural-20);
  color: var(--ic-action-default);
  cursor: auto;
  box-shadow: none;
  outline: none;
}

:host(.in-side-menu) .navigation-group-side-menu:focus {
  outline: var(--ic-hc-focus-outline);
}

:host(.in-side-menu) .navigation-group-side-menu-expanded:hover:not(:focus),
:host(.in-side-menu) .navigation-group-side-menu-collapsed:hover:not(:focus) {
  background-color: var(--ic-action-dark-bg-hover);
  color: var(--ic-action-default);
  cursor: pointer;
}

:host(.in-side-menu) .navigation-group-side-menu-expanded:active:not(:focus),
:host(.in-side-menu) .navigation-group-side-menu-collapsed:active:not(:focus) {
  background-color: var(--ic-action-dark-bg-active);
  color: var(--ic-action-default);
}

:host(.in-side-menu) .navigation-group-side-menu-expanded:focus,
:host(.in-side-menu) .navigation-group-side-menu-collapsed:focus {
  color: var(--ic-action-default);
  box-shadow: var(--ic-border-focus-inset);
  border-radius: var(--ic-border-radius-inset);
  outline: var(--ic-hc-focus-outline);
}

:host .navigation-group-dropdown {
  background-color: var(--ic-architectural-20);
  border-bottom: var(--ic-border-default);
  position: absolute;
  left: 0;
  right: 0;
  padding: var(--ic-space-xs) var(--ic-space-md);
  box-shadow: 0 0.375rem 0.5rem -0.375rem rgba(0 0 0 / 20%);
  z-index: calc(var(--ic-z-index-navigation-item) - 1);
}

:host .navigation-group-dropdown-items-list {
  list-style: none;
  display: flex;
  flex-flow: column wrap;
  align-content: flex-start;
  padding-left: var(--ic-space-md);
  max-height: 16.5rem;
}

:host .chevron-toggle-icon-wrapper {
  display: flex;
  align-items: center;
}

:host .chevron-toggle-icon-wrapper svg {
  transform: rotate(90deg);
  height: var(--ic-space-lg);
  width: var(--ic-space-lg);
}

:host .chevron-toggle-icon-closed svg {
  transform: rotate(-90deg);
}

:host(.expanded) .grouped-links-wrapper {
  height: var(--navigation-child-items-height, auto) !important;
  transition: var(--ic-easing-transition-slow);
  overflow: hidden;
}
:host(.navigation-group-side-nav) .link,
:host(.navigation-group-side-nav) ::slotted(a) {
  height: var(--navigation-child-items-height, auto);
}

:host(.collapsed) .grouped-links-wrapper {
  height: 0;
  transition: var(--ic-easing-transition-slow);
  overflow: hidden;
}

:host(.navigation-group-side-nav) .navigation-group {
  height: var(--navigation-group-height);
  width: var(--navigation-group-width);
  justify-content: var(--navigation-group-justify-content);
  padding-right: var(--navigation-group-expand-toggle-padding);
}

:host(.navigation-group-side-nav) .navigation-group:hover,
:host(.navigation-group-side-nav) .navigation-group:active,
:host(.navigation-group-side-nav) .selected {
  background-color: var(--navigation-group-hover);
  color: var(--navigation-group-text-hover);
}

:host(.navigation-group-side-nav) .navigation-group:focus {
  box-shadow: var(--ic-border-focus-inset);
  border-radius: var(--ic-border-radius-inset);
  background: none;
}

:host(.navigation-group-side-nav) .ic-typography-caption {
  position: var(--navigation-group-title-position);
  left: var(--navigation-group-title-position-left);
  white-space: nowrap;
}
