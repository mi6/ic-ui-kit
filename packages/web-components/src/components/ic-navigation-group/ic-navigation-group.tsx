import {
  Component,
  Element,
  Host,
  Prop,
  State,
  Listen,
  h,
  Method,
} from "@stencil/core";

import {
  DEVICE_SIZES,
  getCurrentDeviceSize,
  getThemeForegroundColor,
  getNavItemParentDetails,
} from "../../utils/helpers";
import { IcNavType, IcTheme } from "../../utils/types";

import chevronIcon from "../../assets/chevron-icon.svg";

@Component({
  tag: "ic-navigation-group",
  styleUrl: "ic-navigation-group.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class NavigationGroup {
  private dropdown: HTMLElement;
  private DYNAMIC_GROUPED_LINKS_HEIGHT_MS = 50;
  private groupEl: HTMLElement;
  private IC_NAVIGATION_ITEM = "ic-navigation-item";
  private mouseGate: boolean = false;
  private nodeName = "IC-NAVIGATION-GROUP";

  @Element() el: HTMLIcNavigationGroupElement;

  @State() deviceSize: number = DEVICE_SIZES.XL;
  @State() dropdownOpen: boolean = false;
  @State() expanded: boolean = true;
  @State() focusStyle = getThemeForegroundColor();
  @State() inTopNavSideMenu: boolean = false;
  @State() navigationType: IcNavType | "";
  @State() parentEl: HTMLElement;

  /**
   *  If `true`, the group will be expandable in the side menu.
   */
  @Prop() expandable: boolean = false;

  /**
   * The label to display on the group.
   */
  @Prop() label: string;

  disconnectedCallback(): void {
    if (this.navigationType === "side") {
      this.parentEl.removeEventListener(
        "sideNavExpanded",
        this.sideNavExpandHandler
      );
    } else if (this.navigationType === "top") {
      this.parentEl.removeEventListener(
        "topNavResized",
        this.topNavResizedHandler
      );
    }
  }

  componentWillLoad(): void {
    this.deviceSize = getCurrentDeviceSize();
    const navParentDetails = getNavItemParentDetails(this.el);
    this.navigationType = navParentDetails.navType;
    this.parentEl = navParentDetails.parent;
    if (this.deviceSize <= DEVICE_SIZES.L && this.navigationType === "top") {
      this.inTopNavSideMenu = true;
    }

    if (this.navigationType === "side") {
      this.parentEl.addEventListener(
        "sideNavExpanded",
        this.sideNavExpandHandler
      );
    } else if (this.navigationType === "top") {
      this.parentEl.addEventListener(
        "topNavResized",
        this.topNavResizedHandler
      );
    }
  }

  componentDidLoad(): void {
    /**
     * debounce is required as the incorrect height was retrieved instantly after
     * componentDidLoad is invoked.
     */
    setTimeout(
      () => this.setInitialGroupedLinksWrapperHeight(),
      this.DYNAMIC_GROUPED_LINKS_HEIGHT_MS
    );
  }

  @Listen("childBlur")
  childBlurHandler(): void {
    this.hideDropdown();
  }

  @Listen("navItemClicked")
  navItemClickHandler(): void {
    this.hideDropdown();
  }

  @Listen("themeChange", { target: "document" })
  themeChangeHandler(ev: CustomEvent): void {
    const theme: IcTheme = ev.detail;
    this.focusStyle = theme.mode;
  }

  /**
   * Sets focus on the nav item.
   */
  @Method()
  async setFocus(): Promise<void> {
    if (this.groupEl) {
      this.groupEl.focus();
    }
  }

  private sideNavExpandHandler = (): void => {
    setTimeout(
      () => this.setInitialGroupedLinksWrapperHeight(),
      this.DYNAMIC_GROUPED_LINKS_HEIGHT_MS
    );
  };

  private topNavResizedHandler = (ev: CustomEvent): void => {
    const newSize = ev.detail.size;
    if (newSize !== this.deviceSize) {
      this.deviceSize = newSize;
      this.inTopNavSideMenu = newSize <= DEVICE_SIZES.L;
    }
  };

  private toggleDropdown() {
    this.dropdownOpen = !this.dropdownOpen;
  }

  private setGroupedNavItemTabIndex = (tabIndexValue: string) => {
    this.el
      .querySelectorAll(this.IC_NAVIGATION_ITEM)
      .forEach((navigationItem) => {
        const navItem =
          navigationItem.shadowRoot.querySelector("a") ||
          navigationItem.querySelector("a");
        if (navItem) {
          navItem.setAttribute("tabindex", tabIndexValue);
        }
      });
  };

  private toggleGroupedLinkWrapperHeight = (
    wrapper: HTMLElement,
    expanded: boolean
  ) => {
    if (wrapper) {
      if (expanded) {
        wrapper.setAttribute(
          "style",
          `--navigation-child-items-height: ${this.getNavigationChildItemsHeight()}`
        );
        this.setGroupedNavItemTabIndex("0");
      } else {
        wrapper.setAttribute("style", `--navigation-child-items-height: 0`);
        this.setGroupedNavItemTabIndex("-1");
      }
    }
  };

  private toggleExpanded = () => {
    this.expanded = !this.expanded;
    const linkWrapper = this.el.shadowRoot.querySelector(
      ".grouped-links-wrapper"
    ) as HTMLElement;
    this.toggleGroupedLinkWrapperHeight(linkWrapper, this.expanded);
  };

  private showDropdown() {
    if (!this.dropdownOpen) {
      this.toggleDropdown();
    }
  }

  private hideDropdown() {
    document.removeEventListener("keydown", this.handleKeydown);
    if (this.dropdownOpen) {
      this.toggleDropdown();
    }
  }

  private handleClick = (ev: MouseEvent) => {
    if (this.navigationType === "top" && ev.detail) {
      this.toggleDropdown();
    } else {
      this.toggleExpanded();
    }
  };

  private handleBlur = (ev: FocusEvent) => {
    const target = ev.relatedTarget as HTMLElement;
    if (!this.el.contains(target)) {
      this.hideDropdown();
    }
  };

  private handleTopNavKeydown = (ev: KeyboardEvent) => {
    if (ev.key === " " || ev.key === "Enter") {
      this.toggleDropdown();
    } else if (!this.inTopNavSideMenu && ev.key === "Escape") {
      this.hideDropdown();
    }
  };

  private handleKeydown = (ev: KeyboardEvent) => {
    if (ev.key === "Enter" || ev.key === " " || ev.key === "Escape") {
      switch (this.navigationType) {
        case "top":
          this.handleTopNavKeydown(ev as KeyboardEvent);
          break;
        case "side":
          this.toggleExpanded();
          ev.preventDefault();
          break;
        default:
          this.toggleExpanded();
          break;
      }
    }
  };

  private handleMouseLeave = (ev: MouseEvent) => {
    const relTarget = ev.relatedTarget as HTMLElement;

    this.mouseGate = false;

    if (
      !this.el.contains(relTarget) &&
      relTarget !== this.dropdown &&
      document.activeElement !== this.el &&
      !this.el.contains(document.activeElement) &&
      relTarget.nodeName === this.nodeName &&
      this.dropdownOpen === true
    ) {
      this.mouseGate = true;
      this.hideDropdown();
    } else if (
      !this.el.contains(relTarget) &&
      relTarget !== this.dropdown &&
      !this.el.contains(document.activeElement)
    ) {
      this.mouseGate = false;
      setTimeout(() => {
        this.dropdownOpen ? this.hideDropdown() : null;
      }, 500);
    }
  };

  private handleMouseEnter = (ev: MouseEvent) => {
    const relTarget = ev.relatedTarget as HTMLElement;
    document.addEventListener("keydown", this.handleKeydown);

    if (relTarget.nodeName === this.nodeName && this.mouseGate === true) {
      this.showDropdown();
    } else if (
      this.dropdownOpen === false &&
      relTarget !== null &&
      this.mouseGate === false
    ) {
      this.mouseGate = true;
      setTimeout(() => {
        this.mouseGate && this.showDropdown();
      }, 500);
    }
  };

  private renderDropdownGroupedLinks = (): HTMLDivElement => (
    <div
      class={{
        ["navigation-group-dropdown"]: !this.inTopNavSideMenu,
        ["navigation-group-dropdown-side-menu"]: this.inTopNavSideMenu,
        ["selected"]: this.dropdownOpen && !this.inTopNavSideMenu,
      }}
      onMouseLeave={!this.inTopNavSideMenu ? this.handleMouseLeave : null}
      ref={(el) => (this.dropdown = el)}
    >
      <nav
        class={{
          ["navigation-group-dropdown-items"]: !this.inTopNavSideMenu,
        }}
      >
        <ul>
          <slot></slot>
        </ul>
      </nav>
    </div>
  );

  private renderGroupedLinks = (): HTMLDivElement => (
    <ul class="grouped-links-wrapper">
      <slot></slot>
    </ul>
  );

  /**
   * Gets the total height of navigation links to improve
   * smoothness of expand/collapse animations
   */
  private getNavigationChildItemsHeight = (): string => {
    const navigationItemLength =
      this.el.querySelectorAll(this.IC_NAVIGATION_ITEM).length || 0;
    const navigationItem =
      this.el.querySelector(this.IC_NAVIGATION_ITEM) || null;

    if (
      this.navigationType === "side" &&
      !!navigationItemLength &&
      navigationItem
    ) {
      return `${
        parseInt(getComputedStyle(navigationItem).height) * navigationItemLength
      }px`;
    }

    return "auto";
  };

  private setInitialGroupedLinksWrapperHeight = () => {
    const linkWrapper = this.el.shadowRoot.querySelector(
      ".grouped-links-wrapper"
    ) as HTMLElement;

    if (linkWrapper && this.expanded) {
      linkWrapper.setAttribute(
        "style",
        `--navigation-child-items-height: ${this.getNavigationChildItemsHeight()}`
      );
    }
  };

  private renderNavigationItems = (): HTMLDivElement | null => {
    if (this.dropdownOpen || (this.inTopNavSideMenu && !this.expandable)) {
      return this.renderDropdownGroupedLinks();
    }

    if (this.navigationType !== "top") {
      return this.renderGroupedLinks();
    }

    return null;
  };

  render() {
    const { label, dropdownOpen, inTopNavSideMenu, expandable } = this;

    const NavigationGroupElement =
      !inTopNavSideMenu || expandable ? "button" : "div";

    return (
      <Host
        class={{
          ["in-side-menu"]: inTopNavSideMenu,
          expanded: this.expanded,
          collapsed: !this.expanded,
          ["navigation-group-side-nav"]: this.navigationType === "side",
        }}
        role="listitem"
      >
        <NavigationGroupElement
          onMouseEnter={
            !inTopNavSideMenu &&
            this.navigationType === "top" &&
            this.handleMouseEnter
          }
          onMouseLeave={this.navigationType === "top" && this.handleMouseLeave}
          tabindex={inTopNavSideMenu && !expandable ? "-1" : "0"}
          onBlur={this.handleBlur}
          onClick={expandable ? this.handleClick : null}
          onKeyDown={this.handleKeydown}
          class={{
            ["navigation-group"]: true,
            [this.focusStyle]: !inTopNavSideMenu,
            ["navigation-group-side-menu"]: inTopNavSideMenu && !expandable,
            ["navigation-group-side-menu-collapsed"]:
              inTopNavSideMenu && expandable && !dropdownOpen,
            ["navigation-group-side-menu-expanded"]:
              inTopNavSideMenu && expandable && dropdownOpen,
            ["selected"]: dropdownOpen && !inTopNavSideMenu,
          }}
          ref={(el) => (this.groupEl = el)}
          aria-expanded={dropdownOpen || this.expanded ? "true" : "false"}
          aria-haspopup={
            !inTopNavSideMenu && this.navigationType === "top"
              ? "true"
              : "false"
          }
        >
          <ic-typography
            variant={this.navigationType === "side" ? "caption" : "label"}
          >
            {label}
          </ic-typography>
          {this.navigationType === "side" && expandable && (
            <div
              class={{
                "chevron-toggle-icon-wrapper": true,
                "chevron-toggle-icon-closed": this.expanded,
              }}
              innerHTML={chevronIcon}
            ></div>
          )}
        </NavigationGroupElement>
        {this.renderNavigationItems()}
      </Host>
    );
  }
}
