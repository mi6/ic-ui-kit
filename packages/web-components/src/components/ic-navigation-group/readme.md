# ic-navigation-group



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute    | Description                                               | Type      | Default     |
| ------------ | ------------ | --------------------------------------------------------- | --------- | ----------- |
| `expandable` | `expandable` | If `true`, the group will be expandable in the side menu. | `boolean` | `false`     |
| `label`      | `label`      | The label to display on the group.                        | `string`  | `undefined` |


## Methods

### `setFocus() => Promise<void>`

Sets focus on the nav item.

#### Returns

Type: `Promise<void>`




## CSS Custom Properties

| Name                           | Description                      |
| ------------------------------ | -------------------------------- |
| `--ic-z-index-navigation-item` | z-index of navigation group item |


## Dependencies

### Depends on

- [ic-typography](../ic-typography)

### Graph
```mermaid
graph TD;
  ic-navigation-group --> ic-typography
  style ic-navigation-group fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


