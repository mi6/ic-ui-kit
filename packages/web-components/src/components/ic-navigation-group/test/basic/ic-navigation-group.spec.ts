import { newSpecPage } from "@stencil/core/testing";
import { waitForNavGroupLoad } from "../../../../testspec.setup";
import { NavigationItem } from "../../../ic-navigation-item/ic-navigation-item";
import { SideNavigation } from "../../../ic-side-navigation/ic-side-navigation";
import { TopNavigation } from "../../../ic-top-navigation/ic-top-navigation";
import { DEVICE_SIZES } from "../../../../utils/helpers";
import { NavigationGroup } from "./../../ic-navigation-group";
import * as helpers from "../../../../utils/helpers";
import { waitForTimeout } from "../../../../testspec.setup";

const ev = {
  relatedTarget: {
    nodeName: "IC-NAVIGATION-GROUP",
  },
  target: {
    nodeName: "IC-NAVIGATION-GROUP",
  },
};

const timeOut = 1000;

describe("ic-navigation-group", () => {
  it("should render with label", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup],
      html: `<ic-navigation-group label="Group label"></ic-navigation-group>`,
    });
    await waitForNavGroupLoad();
    expect(page.root).toMatchSnapshot("renders-with-label");
  });

  it("should test handleMouseEnter", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup],
      html: `<ic-navigation-group label="Group label"></ic-navigation-group>`,
    });
    await waitForNavGroupLoad();
    page.rootInstance.mouseGate = true;
    expect(page.rootInstance.dropdownOpen).toBe(false);
    await page.rootInstance.handleMouseEnter(ev);
    await waitForTimeout(timeOut);
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(true);
    await page.rootInstance.hideDropdown();
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(false);

    ev.relatedTarget.nodeName = "ANOTHER-NAME";
    page.rootInstance.mouseGate = false;
    await page.rootInstance.handleMouseEnter(ev);
    await waitForTimeout(timeOut);
    await page.waitForChanges();
    expect(page.rootInstance.mouseGate).toBe(true);

    jest.spyOn(document, "addEventListener");

    await page.rootInstance.handleMouseEnter(ev);
    await waitForTimeout(timeOut);
    await page.waitForChanges();

    expect(document.addEventListener).toHaveBeenCalledTimes(1);

    await page.rootInstance.showDropdown();
    await page.waitForChanges();

    await page.rootInstance.childBlurHandler();
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(false);

    await page.rootInstance.showDropdown();
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(true);

    await page.rootInstance.navItemClickHandler();
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(false);
    ev.relatedTarget.nodeName = "IC-NAVIGATION-GROUP";
  });

  it("should test toggleExpanded", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup, NavigationItem],
      html: `<ic-navigation-group label="Group label">
        <ic-navigation-item href="/" label="Home"></ic-navigation-item>
      </ic-navigation-group>`,
    });
    await waitForNavGroupLoad();

    expect(page.rootInstance.expanded).toBe(true);
    await page.rootInstance.toggleExpanded();
    await page.waitForChanges();
    expect(page.rootInstance.expanded).toBe(false);
    await page.rootInstance.toggleExpanded();
    await page.waitForChanges();
    expect(page.rootInstance.expanded).toBe(true);
  });

  it("should test click handler", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup, NavigationItem],
      html: `<ic-navigation-group label="Group label">
        <ic-navigation-item href="/" label="Home"></ic-navigation-item>
      </ic-navigation-group>`,
    });
    await waitForNavGroupLoad();

    expect(page.rootInstance.expanded).toBe(true);
    await page.rootInstance.handleClick();
    await page.waitForChanges();
    expect(page.rootInstance.expanded).toBe(false);

    expect(page.rootInstance.dropdownOpen).toBe(false);
    page.rootInstance.navigationType = "top";
    await page.rootInstance.handleClick({ detail: { something: "set" } });
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(true);
  });

  it("should test top nav key down handler", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup, NavigationItem],
      html: `<ic-navigation-group label="Group label">
        <ic-navigation-item href="/" label="Home"></ic-navigation-item>
      </ic-navigation-group>`,
    });
    await waitForNavGroupLoad();

    expect(page.rootInstance.dropdownOpen).toBe(false);
    await page.rootInstance.showDropdown();
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(true);
    await page.rootInstance.handleTopNavKeydown({ key: "Escape" });
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(false);
    await page.rootInstance.handleTopNavKeydown({ key: "Enter" });
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(true);
  });

  it("should test handleBlur function", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup, NavigationItem],
      html: `<ic-navigation-group label="Group label">
      <ic-navigation-item href="/" label="Home"></ic-navigation-item>
      <ic-navigation-item href="/" label="andAway"></ic-navigation-item>
      <ic-navigation-item href="/" label="closerEachDay"></ic-navigation-item>
      </ic-navigation-group>`,
    });

    await waitForNavGroupLoad();
    await page.rootInstance.showDropdown();
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(true);
    await page.rootInstance.handleBlur(ev);
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(false);
  });

  it("should test handleMouseLeave function", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup, NavigationItem],
      html: `<ic-navigation-group label="Group label">
        <ic-navigation-item href="/" label="Home"></ic-navigation-item>
        <ic-navigation-item href="/" label="andAway"></ic-navigation-item>
      </ic-navigation-group>`,
    });
    await waitForNavGroupLoad();
    ev.target.nodeName = "IC-NOT-NAV-GROUP";

    await page.rootInstance.showDropdown();
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(true);
    await page.rootInstance.handleMouseLeave(ev);
    await waitForTimeout(600);
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(false);
    // check other branch of 'if'
    ev.relatedTarget.nodeName = "IC-NOT-NAV-GROUP";
    await page.rootInstance.showDropdown();
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(true);
    await page.rootInstance.handleMouseLeave(ev);
    await waitForTimeout(timeOut);
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(false);
    // check if null value called
    await page.rootInstance.handleMouseLeave(ev);
    await waitForTimeout(timeOut);
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(false);
    ev.relatedTarget.nodeName = "IC-NAVIGATION-GROUP";
    ev.target.nodeName = "IC-NAVIGATION-GROUP";
  });

  it("should test key down handler", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup, NavigationItem],
      html: `<ic-navigation-group label="Group label">
        <ic-navigation-item href="/" label="Home"></ic-navigation-item>
      </ic-navigation-group>`,
    });
    await waitForNavGroupLoad();

    expect(page.rootInstance.expanded).toBe(true);
    await page.rootInstance.handleKeydown({ key: "Enter" });
    await page.waitForChanges();
    expect(page.rootInstance.expanded).toBe(false);

    page.rootInstance.navigationType = "side";
    await page.rootInstance.handleKeydown({
      key: " ",
      preventDefault: (): void => null,
    });
    await page.waitForChanges();
    await waitForTimeout(600);
    expect(page.rootInstance.expanded).toBe(true);

    expect(page.rootInstance.dropdownOpen).toBe(false);
    page.rootInstance.navigationType = "top";
    await page.waitForChanges();
    await page.rootInstance.handleKeydown({ key: "Enter" });
    await page.waitForChanges();
    expect(page.rootInstance.dropdownOpen).toBe(true);
  });

  it("should test theme change", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup],
      html: `<ic-navigation-group label="Group label"></ic-navigation-group>`,
    });
    await waitForNavGroupLoad();
    await page.rootInstance.themeChangeHandler({ detail: { mode: "dark" } });
    await page.waitForChanges();
    expect(page.rootInstance.focusStyle).toBe("dark");

    await page.rootInstance.themeChangeHandler({ detail: { mode: "light" } });
    await page.waitForChanges();
    expect(page.rootInstance.focusStyle).toBe("light");
  });

  it("should test inside side navigation", async () => {
    const page = await newSpecPage({
      components: [SideNavigation, NavigationGroup, NavigationItem],
      html: `<ic-side-navigation version="v0.0.0" app-title="ACME" status="BETA" collapsed-icon-labels="true">
      <ic-navigation-group
            slot="primary-navigation"
            label="Second navigation group"
            expandable="true"
          >
        <ic-navigation-item href="/" label="Home"></ic-navigation-item>
      </ic-navigation-group>
    </ic-side-navigation>`,
    });
    await waitForNavGroupLoad();

    // test disconnected callback
    page.setContent("");
  });

  it("should test top nav resize handler", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup],
      html: `<ic-navigation-group label="Group label"></ic-navigation-group>`,
    });
    await waitForNavGroupLoad();

    await page.rootInstance.topNavResizedHandler({
      detail: { size: DEVICE_SIZES.S },
    });
    await page.waitForChanges();
    expect(page.rootInstance.deviceSize).toBe(DEVICE_SIZES.S);
    expect(page.rootInstance.inTopNavSideMenu).toBe(true);
  });

  it("should test side nav expand handler", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup],
      html: `<ic-navigation-group label="Group label"></ic-navigation-group>`,
    });
    await waitForNavGroupLoad();
    await page.rootInstance.sideNavExpandHandler();
    await page.waitForChanges();
    await waitForNavGroupLoad();
    const style = page.root.shadowRoot
      .querySelector(".grouped-links-wrapper")
      .getAttribute("style");
    expect(style).toBe("--navigation-child-items-height: auto;");
  });

  it("should test setting nav item tab index", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup, NavigationItem],
      html: `<ic-navigation-group
            label="Navigation group"
          >
        <ic-navigation-item href="/" label="Home"></ic-navigation-item>
      </ic-navigation-group>`,
    });
    await waitForNavGroupLoad();

    let navItemLink = page.root
      .querySelector("ic-navigation-item")
      .shadowRoot.querySelector("a");
    expect(navItemLink.tabIndex).toBe(-1);

    await page.rootInstance.setGroupedNavItemTabIndex("0");
    await page.waitForChanges();
    navItemLink = page.root
      .querySelector("ic-navigation-item")
      .shadowRoot.querySelector("a");

    expect(navItemLink.tabIndex).toBe(0);
  });

  it("should call other methods ", async () => {
    const page = await newSpecPage({
      components: [NavigationGroup],
      html: `<ic-navigation-group label="Group label"></ic-navigation-group>`,
    });
    await waitForNavGroupLoad();
    await page.root.setFocus();
  });

  // NOTE: This must go last as mocks getCurrentDeviceSize function, which will apply to all subsequent tests
  it("should test inside top navigation", async () => {
    const myfunc = jest.fn().mockReturnValue(DEVICE_SIZES.S);
    Object.defineProperty(helpers, "getCurrentDeviceSize", {
      value: myfunc,
    });

    const page = await newSpecPage({
      components: [TopNavigation, NavigationGroup, NavigationItem],
      html: `<ic-top-navigation
        app-title="ApplicationName"
      >
        <ic-navigation-group
          slot="navigation"
          label="Navigation group"
          expandable="true"
        >
          <ic-navigation-item label="Navigation 1" href="/"></ic-navigation-item>
        </ic-navigation-group>
      </ic-top-navigation>`,
    });
    await waitForNavGroupLoad();

    // test disconnected callback
    page.setContent("");
  });
});
