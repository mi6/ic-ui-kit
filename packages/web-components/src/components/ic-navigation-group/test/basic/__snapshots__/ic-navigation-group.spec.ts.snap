// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-navigation-group should render with label: renders-with-label 1`] = `
<ic-navigation-group class="expanded" label="Group label" role="listitem">
  <mock:shadow-root>
    <button aria-expanded="true" aria-haspopup="false" class="light navigation-group" tabindex="0">
      <ic-typography variant="label">
        Group label
      </ic-typography>
    </button>
    <ul class="grouped-links-wrapper" style="--navigation-child-items-height: auto;">
      <slot></slot>
    </ul>
  </mock:shadow-root>
</ic-navigation-group>
`;
